Bug 95 firmware 1.0.4 error - a.k.a. - adding left/right black shields

rolkar opened the pull request:
Hi, the current pull request is WIP (Work In Progress). So - it shall not be merged.

The solution seems to work.

In the x3f_process.c file there is a debug printout (search for XXXX) that writes the current mean values for the four areas. They are VERY similar - so that is good. Very good. The new left/right areas give the same value as the top/bottom. Nice!

In the x3f_image.c file is the heart of the solution in the function x3f_get_fake_camf_matrix. When getting the two Fake matrix names, it instead fetches the value from the DarkShieldColRange and constricts the matrices. Then, nearly no code needs to be touched, Is that a good solution? I tried to find a better, but could not. But no, it sux. Lets ignore that for now :)

I have two TODO in that function. I have no idea wether the image->rows is the correct value to set. Or - if you need to take keep_area into account.

So - that is it. Please take a look and see if you like it and if you believe in it.

As I said, it make beautiful pictures from the image that previouslu crashed.

erikrk commented on src/x3f_image.c:
```diff
@@ -115,9 +144,17 @@
+    if (!x3f_get_camf_matrix(x3f, name, 4, 0, 0, M_UINT, rect))
+      return 0;
+  }
+
```
To make it a bit less dirty, I would rather use a lookup table with names of the virtual rectangles with corresponding function pointers here.

rolkar replied:
A thought ... but even that would be a kludge.

I first started to do something in the get_black_level function, where I would call different get_rect functions and then use the rects as arguments. Then there would be nu ugliness in the library functions. But, then I could not reuse your thinking regarding the keep_areas and mappings from rects to crops. I then had to reinvent all that stuff.

erikrk replied:
What you could do is refactor x3f_get_camf_rect, moving the transformation code (line 155 and onwards) to a separate function that takes a rectangle as a parameter. That function could then by used both by x3f_get_camf_rect and a new function that obtains rectangles from column ranges.

rolkar replied:
Good idea, then this Fake thing can be history!

erikrk commented on src/x3f_process.c:
```diff
@@ -69,20 +59,49 @@ static int get_black_level(x3f_t *x3f,
+	x3f_printf(WARN, "XXXX pixels = %d mean[%d] = %f\n",
+		   pixels,
+		   color,
+		   (double)black_sum[color]/pixels);
```
This doesn't print mean for each area but rather cumulative means.

rolkar replied:
Yeah, I know. But still ... it is very stable. It is virtually the same mean value all over.

erikrk commented on src/x3f_image.c:
```diff
@@ -98,6 +99,34 @@
+    rect[0] = column[2];
+    rect[1] = 0;
+    rect[2] = column[3];
+    rect[3] = image->rows; /* TODO - is this right? */
```
The rows have to be obtained from KeepImageArea.

rolkar replied:
Both the min and the max? I.e. 1 and 3.
And - what about 0 and 2, the columns?
keep_area all the way? Or is the columns in the meta data already in keep_area coordinates?

erikrk replied:
Actually, 1 and 3 don't matter too much, since the rectangle will be clipped to KeepImageArea subsequently in x3f_get_camf_rect. I would just set them to 0 and UINT32_MAX, so that x3f_get_fake_camf_matrix doesn't need to have unnecessary knowledge.

I think 0 and 2 should be taken directly from DarkShieldColRange, without modification.

rolkar replied:
OK - good idea with UINT32_MAX. Ill do that. And write a comment something like your text above.

rolkar commented:
I have now fixed it so that we do not need the kludge to deep in the library know about the fake crop rectangles, just as Erik recommended.

I would say that it might be ready for merge now.

There is only one fly in the ointment. Looking at lots of X3F files, there are some few where one of the areas gives a totally too high black level. I mean, totally to high, like 1000 when it should be 7. I assume this shielded area is broken somehow. To solve this, we need a better algorithm for choosing the black level. Today we sum up all areas before calculating the black level. That is not good enough.

But, I assume this is another issue and do not stop us from merging.

One info - I have three sd Quattro H files. In two of them the top shield seems broken. Maybe just a coincidence, as all three images are from the same camera.

Still, I recommend that we file a new issue instaed of delaying this merge.

erikrk commented on src/x3f_image.c:
```diff
@@ -166,16 +168,56 @@
+  rect[1] = 0;           /* Cropped automatically later */
+  rect[3] = UINT32_MAX;  /* Cropped automatically later */
+
+  if (!strcmp(name, "FakeDarkShieldLeft")) {
```
I don't think that this Fake* nomenclature should be used any more. x3f_crop_area_column should really take the name of a list of column ranges, like DarkShieldColRanges, and an index instead.

rolkar replied:
I agree on the fake thing. Ill remove it. But ... the function is so special, so it do not need the DarkShieldColRanges parameter. It can be implicit.

Then, I am a bit spoliled by languages with atoms, and I am 100% fed up with inventing defines for left and right, and I really do not fancy sending 0 and 1. The strings "LEFT" and "RIGHT" would just feel nice :) OK - then you need to use strcmp. je je ... ill add two global defines for COL_LEFT_RECT and COL_RIGHT_RECT or some such. But ... I do not like it. Not at all.

erikrk replied:
Yes, C is very primitive in this regard. I doesn't have atoms, and it doesn't really have sting manipulation either. All it has it sting constants. The rest is handled by th library.

I don't think indexes are really wrong here though. It's after all just a list of column rages. There is no guarantee that one of them will be to the left and the other to the right, or even that it will be two of them.

erikrk commented on src/x3f_image.c:
```diff
@@ -14,6 +14,7 @@
 
 #include <stdio.h>
 #include <assert.h>
+#include <string.h>
```
This should not be needed.

erikrk commented on src/x3f_process.c:
```diff
@@ -69,20 +59,57 @@ static int get_black_level(x3f_t *x3f,
+      for (color = 0; color < colors; color++)
+	x3f_printf(INFO, "    mean[%d] = %f\n",
+		   color,
+		   (double)black_sum[color]/pixels);
```
These printouts have to be removed before merging, or at least changed to DEBUG. If they are to be kept (as DEBUG), I also think they should be improved so that they print the mean for each area instead of cumulative means.

rolkar replied:
Ill make it DEBUG. We will need this to debug the too high value anomaly.

erikrk commented on src/x3f_process.c:
```diff
@@ -91,11 +118,12 @@ static int get_black_level(x3f_t *x3f,
+  for (i=0; i<4; i++)
+    if (use[i]) 
+      pixels += sum_area_sqdev(x3f, area[i], colors,
+			       black_level, black_sum_sqdev);
```
In case debug printouts are to be kept, standard deviations should really be printed too.

rolkar replied:
OK - ill add the stdd also. It will help debugging the too high value.

rolkar commented:
I have now fixed all the above.
The "get column" do now take an enum to decide left or right.m I think that is OK. If we, in the future, finds some multi column thingie with more than two areas, then we will add an index instead.

I have fixed the printout so that it no longer writes the accumulated value and also writes deviation.
Doing so, I had to no longer accumulate values along the way, so now each call calculates its own.

Then, sometimes, one of the areas is dead wrong.
Then, the total sum black level is way off.
And then, the deviation is also way off, as it is relative to the black level.
The deviation should really be calculated per call.

I think it would be nice to close this issue now anyhow.
I think fixing a more elaborate choice of black level and all
that comes with that is another story.

NOTE - in the last commit, the black level printouts still are
written to INFO. Much easier to debug then.
I will, of course, move it to DEBUG before any merge.

rolkar commented:
And now I have change the printouts to DEBUG.

erikrk commented on src/x3f_process.c:
```diff
@@ -22,15 +22,12 @@
-static int sum_area(x3f_t *x3f, char *name,
-		    x3f_area16_t *image, int rescale, int colors,
-		    uint64_t *sum /* in/out */)
+static int sum_area(x3f_t *x3f, x3f_area16_t area, int colors,
```
x3f is an unused parameter here, and also in sum_area_sqdev.

erikrk commented:
Otherwise, I think it looks good now. One question though: Have you run "make check"? Don't the hashes need updating since you get somewhat different results when all four areas are used for calculating black level.

rolkar commented:
Ouch!

Yes, the trillion of hashes are incorrect.

Any way of updating them automatically?

/Roland



On 2017-05-06 21:02, Erik Karlsson wrote:
>
> Otherwise, I think it looks good now. One question though: Have you 
> run "make check"? Don't the hashes need updating since you get 
> somewhat different results when all four areas are used for 
> calculating black level.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Kalpanika/x3f/pull/99#issuecomment-299659649>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAVzmFQyx4qIkVcYsuYni_9sj9V8oKg2ks5r3MPggaJpZM4NRFQ3>.
>



---
Detta e-postmeddelande har sökts igenom efter virus med antivirusprogram från Avast.
https://www.avast.com/antivirus

erikrk commented on src/x3f_process.c:
```diff
@@ -66,40 +60,111 @@ static int get_black_level(x3f_t *x3f,
+      for (color = 0; color < colors; color++) {
+	x3f_printf(DEBUG, "    dev[%d] = %g\n",
+		   color,
+		   sqrt(black_sqdev[color]/pixels));
```
As mentioned, this printout is incorrect, since it uses the overall mean to calculate the standard deviation for individual areas. This need to ether be fixed or removed. In the current state it more confusing than useful. To fix it, you would av do do an additional call too sum_area_squdev in the first loop that calculates the means. The result of this call would be used only for printouts.

rolkar replied:
Removed

mmroden commented:
I can update the hashes if we all agree that the results are what they should be; if the black level is off enough to make the conversion bad and make the hash wrong, then the hash has done its job of making us check the output.

On May 6, 2017, 12:26 PM -0700, Erik Karlsson <notifications@github.com>, wrote:
>
> @erikrk commented on this pull request.
>
>
> In src/x3f_process.c (https://github.com/Kalpanika/x3f/pull/99#discussion_r115127193):
>
> > + black_sqdev_sum = alloca(colors*sizeof(double)); + for (i=0; i<colors; i++) black_sqdev_sum[i] = 0.0; + + for (i=0; i<4; i++) + if (use[i]) { + int color; + int pixels = sum_area_sqdev(area[i], colors, black_level, black_sqdev); + + pixels_sum += pixels; + + x3f_printf(DEBUG, " %s (%d)\n", name[i], pixels); + + for (color = 0; color < colors; color++) { + x3f_printf(DEBUG, " dev[%d] = %g\n", + color, + sqrt(black_sqdev[color]/pixels));
>
> As mentioned, this printout is incorrect, since it uses the overall mean to calculate the standard deviation for individual areas. This need to ether be fixed or removed. In the current state it more confusing than useful. To fix it, you would av do do an additional call too sum_area_squdev in the first loop that calculates the means. The result of this call would be used only for printouts.
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub (https://github.com/Kalpanika/x3f/pull/99#pullrequestreview-36650696), or mute the thread (https://github.com/notifications/unsubscribe-auth/AAHGG7n7rBRf7YBXC8Y9gzE-qAhuq9CEks5r3MlZgaJpZM4NRFQ3).
>
>
>

rolkar commented:
Thanks Marc.

I will fix the "last" comment by Erik :) It is easy, but it is late and I am tired. So, I will probably do it tomorrow morning.

One thought: I have been looking at several pictures, and they look fine. And, we now will have a new update, hopefully soon, that makes a better decision on black level. It might then be a waste of time to update all hashes now. Maybe wait until after that fix?

rolkar commented:
I fixed the output now.

So, if we choose to not update the image hashes, then this is a merge candidate.

rolkar commented:
As this change only affects the black level, and we are going to refine the black level computation anyhow, I do not think it is a bad thing to merge this. So, I do it.
